### PR TITLE
Enable unicode characters for TS channel and Arma server names

### LINKED ISSUE #579: Non-ASCII characters break channel switching
**Arma 3 Version:** `1.84` stable
**CBA Version:** `3.8.0` stable
**ACRE2 Version:** `2.6.1` stable

**Mods:**
```
- CBA_A3
- acre2
```

**Description:**
Non-ASCII characters in the server name breaks the channel switching system.

**Steps to reproduce:**
- [Sample server.cfg](https://files.slack.com/files-pri/T0J06UTRP-FCNEBG5A7/download/server.cfg)
- `hostName = "7.Kolordu Komutanlığı Teamspeak (ACRE) | 7KolorduTS | ";`

**Where did the issue occur?**
- Dedicated

**Placed Modules:**
- None

Issue created for max315 from Slack.

### PR DESCRIPTION
**Pull request by tbeswick96**
**When merged this pull request will:**
- Allow unicode characters in TeamSpeak channel names and Arma server names
- Change relevant TS3 methods for channel switching to use `wstring` instead of `string`
- Add 2 conversion methods to convert between `string` and `wstring`
- Stop the `TextMessage` `__isascii` check from running only for `setTs3ChannelDetails` procedure
- Fix #579 
- Tested locally using a TeamSpeak channel name containing `Grę` and a local Arma MP instance with server name `Grę`. Channel switching works as expected. All other functionality seems unaffected (Could use a smoke test if anyone is interested in building and testing)

---

Notes:
- The data coming from arma comes in as `char*`. This gets interpreted as ASCII, and the majority of data passed around inside ACRE is ASCII.
- The unicode conversion for channel switching only happens in the RPC function `setTs3ChannelDetails`. This means data is sent into it in the "wrong" format, is converted and used accordingly, and where needed passed back out in the "wrong" format. The out I think is only for channel passwords. So it may be the case that a channel password containing unicode doesn't work as expected, however the bytes remain the same so TS should interpret it correctly. I've not tested this.

### CODE REVIEW

**Comment by TheMagnetar** [comment]
What about using `std::string test = u8"test"` instead of wstring?
https://en.cppreference.com/w/cpp/language/string_literal

**Comment by tbeswick96** [comment]
> What about using `std::string test = u8"test"` instead of wstring?
> [en.cppreference.com/w/cpp/language/string_literal](https://en.cppreference.com/w/cpp/language/string_literal)

u8, as far as i can see, can only be used for literals, so that doesn't help much for getting the string from a variable

File: extensions/src/ACRE2Shared/StringConversions.cpp
Comment (TheMagnetar) [inline extensions/src/ACRE2Shared/StringConversions.cpp]: Fixed integer sizes
```suggestion
    int32_t num_chars = MultiByteToWideChar(CP_UTF8, 0, str.c_str(), str.length(), NULL, 0);
```

File: extensions/src/ACRE2Shared/TextMessage.cpp
Comment (TheMagnetar) [inline extensions/src/ACRE2Shared/TextMessage.cpp]:
```suggestion
        for (size_t x = 0U; x < len; x++) {
```

**Review by TheMagnetar — CHANGES_REQUESTED** [review]
Great stuff! only minor changes. The PR looks good but I would like to see how the mumble plugin supports it, before merging this branch into master.

File: extensions/src/ACRE2Shared/TextMessage.cpp
Comment (tbeswick96) [reply extensions/src/ACRE2Shared/TextMessage.cpp]: This line wasn't edited by me but have made the change anyway

**Comment by tbeswick96** [comment]
@TheMagnetar Thanks, I don't have Mumble though so I can't test this with that, someone else with that all setup will need to lend a hand with that

**Comment by TheMagnetar** [comment]
For sure, this is in our hands atm. Thanks for the PR @tbeswick96

**Comment by jonpas** [comment]
Mumble compatibility branch here: https://github.com/IDI-Systems/acre2/tree/mumble-unicode

Tried with `Test lığı` as server name and it switched correctly on Mumble. However on TeamSpeak it moves me to the wrong channel (just uses first one that contains "ACRE" it seems). Could use another set of eyes.

**Comment by jonpas** [comment]
Did anyone get a chance to test this with my example? @tbeswick96

**Comment by tbeswick96** [comment]
I've not forgotten about this, just really busy. Intending to try and get this done in a few weeks

**Comment by jonpas** [comment]
Most likely replaced by #1326. Please reopen if not the case.

**Pull request closed without merging**
